source/administration/data-retention.rst

```diff
@@ -0,0 +1,68 @@
+
+By default, Mattermost stores all message history providing an unlimited search history to admins and end users.
+
+In Enterprise Edition E20, you may set a custom policy for how long messages and file uploads are kept in Mattermost channels and direct messages.
+
+.. warning:: Once a message or a file is deleted, the action is irreversible. Please be careful when setting up a custom data retention policy.
+
+.. toctree::
+    :maxdepth: 2
+
+Set Up Guide
+--------------
+
+To set a custom data retention policy:
+
+1. Go to **System Console > Advanced > Data Retention Policy (Beta)**
+2. Set your desired retention policy for messages. Messages, including file attachments, older than the duration you set will be deleted nightly. The minimum message retention time is one day.
+3. Similarly, set your desired retention policy for file uploads. File uploads older than the duration you set will be deleted nightly from your file storage system, either from your local disk or your Amazon S3 service as specified in **System Console > Files > Storage**. The minimum file retention time is one day.
+4. Set the start time of the daily scheduled data retention job. Choose a time when fewer people are using your system. Must be a 24-hour time stamp in the form HH:MM.
+
+Save the settings and restart your server. You're now all set! Data older than the duration you set are now deleted daily at the specified server time.
+
+You may also run the deletion job manually anytime by clicking **Run Deletion Job Now** button in **System Console > Advanced > Data Retention Policy (Beta)**.
+
+Frequently Asked Questions (FAQ)
+---------------------------------
+
+What happens when a message is deleted?
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The message is removed from the Mattermost user interface and deleted from the Posts table. The message is no longer searchable and cannot be retrieved in pinned posts or flagged posts lists.
+
+Replies that did not exceed the message duration are still displayed in the user interface. However, further replies to thread are no longer permitted.
+
+What happens when a file is deleted?
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The file attachment is removed from the Mattermost user interface, and deleted from the FileInfo table and from your local disk or Amazon S3 service as specified in **System Console > Files > Storage**.
+
+Why didn't an old file get deleted after running the deletion job?
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+First, check the data deletion was successful in the deletion job table in **System Console > Advanced > Data Retention Policy (Beta)**.
+
+Second, if the files were uploaded prior to Mattermost v4.2, you will need to delete the files manually from your local file storage or Amazon S3 storage. You may do this by simply deleting the relevant folders. // XXX George, any other info/help we can provide?
```

Comment:
So prior to 4.2 the folder structure was something like `teamid/channelid/fileid/...` so the only way to delete files is either to figure out which one and manually delete the right bits, or to just delete all these folders, removing all files from pre 4.2. My advice is to wait until the data retention period you want gives all of pre-4.2 data to be past the cutoff, and then to manually delete the whole lot. In practice, I think there will be very few if any E20 customers who have pre 4.2 systems who apply data retention policy.

Reply:
We already know of a few customers who have run Mattermost prior to v4.2 and are interested in this feature, so I think a few will hit this issue.

That being said, is the recommendation to:
 1. enable file retention policy
 2. wait until all files uploaded to the folder `teamid/channelid/fileid/...` fall into the data retention policy duration
 3. manually delete all these folders

Is there a way for the customer to know when the appropriate timing for 2 is? Like a particular time field in the Files table? Since they may not remember when they upgraded to v4.2.

Reply:
The 3 steps are correct. I think it's actually even simpler: once the appropriate time has elapsed they can just delete the `teams/` folder in the root of their Mattermost storage directory/bucket. I'll check with HH to make sure there's nothing else in that folder, but I'm pretty sure there isn't.

As for *when* they can get away with deleting it, I can probably construct a (very slow on large installs) query they can run on a read-replica to estimate that date. But ideally, they'd be able to figure it out (at least roughly) on their own.

Reply:
@grundleborg Updated the FAQ. 

There are some customers with ~ 1 year retention policy (already running Mattermost pre-4.2) so we might need to be prepared to give them extra support on it. But agree the number will be small so we can just handle them on a case-by-case basis